src/forder.c

```diff
@@ -801,6 +800,145 @@ SEXP forder(SEXP DT, SEXP by, SEXP retGrpArg, SEXP sortGroupsArg, SEXP ascArg, S
   return ans;
 }
 
+// all(x==1L)
+static bool all1(SEXP x) {
+  if (!isInteger(x))
+    error("internal error: all1 got non-integer"); // # nocov
+  int *xp = INTEGER(x);
+  for (int i=0; i<LENGTH(x); ++i) {
+    if (xp[i] != 1)
+      return false;
+  }
+  return true;
+}
+
+// identical(cols, head(chmatch(key(x), names(x)), length(cols)))
+bool colsKeyHead(SEXP x, SEXP cols) {
+  if (!isNewList(x))
+    error("internal error: 'x' must be a list"); // # nocov
+  if (!isInteger(cols))
+    error("internal error: 'cols' must be an integer"); // # nocov
+  SEXP key = getAttrib(x, sym_sorted);
+  if (isNull(key) || (length(key) < length(cols)))
+    return false;
+  SEXP names =  getAttrib(x, R_NamesSymbol);
+  SEXP keynames = PROTECT(chmatch(key, names, 0));
+  int *keynamesp = INTEGER(keynames), *colsp = INTEGER(cols);
+  for (int i=0; i<LENGTH(cols); ++i) {
+    if (colsp[i]!=keynamesp[i]) {
+      UNPROTECT(1);
+      return false;
+    }
+  }
+  UNPROTECT(1);
+  return true;
+}
+
+// paste0("__", cols, collapse="")
+SEXP idxName(SEXP cols) {
+  if (!isString(cols))
+    error("internal error: 'cols' must be a character"); // # nocov
+  SEXP char_underscore2 = PROTECT(ScalarString(mkChar("__")));
+  SEXP char_empty = PROTECT(ScalarString(mkChar("")));
+  SEXP sym_paste0 = install("paste0");
+  SEXP call_paste0 = PROTECT(lang4(sym_paste0, char_underscore2, cols, char_empty));
+  SET_TAG(CDDDR(call_paste0), install("collapse"));
+  SEXP ans = PROTECT(eval(call_paste0, R_GlobalEnv));
+  UNPROTECT(4);
+  return ans;
+}
+
+// attr(attr(x, "index"), idxName(names(x)[cols]))
+SEXP getIndex(SEXP x, SEXP cols) {
```

Comment:
we could provide also an R wrapper for this, but needs to handle `cols` character as well, now only integer

Reply:
Follow-up issue: #6262